### PR TITLE
cellPamf implementation

### PR DESCRIPTION
**Pull request by capriots**
Should be fully LLE accurate so I set it to HLE by default and reduced the log level of all functions to notice.

The library supports the older PSMF format used on the PSP, however I could not find any evidence that this functionality is actually used anywhere. The PSP emulator uses a different library (dev_flash/pspemu/release/psmf.sprx).
I implemented PSMF support, but it might be better to just throw an exception for now if CellPamfReader::isPsmf is set since there is a lot of duplicate code, not sure about that.

There are two undocumented functions in libpamf.sprx: cellPamfReaderGetEpWithTimeStamp and cellPamfReaderGetEpWithIndex. The names match the NIDs. No idea if they are used anywhere.

### CODE REVIEW

**Comment by Darkhost1999** [comment]
PS3_GAME/USRDIR/assets/g5/pamf/ubi.pam
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/14152481/RPCS3.log)
No complaints here. Looks good.

**Comment by Megamouse** [comment]
Nice 👍
The review is probably gonna take a while though

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 1]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
 {
-	cellPamf.todo("cellPamfReaderGetEpIteratorWithTimeStamp(pSelf=*0x%x, pTimeStamp=*0x%x, pIt=*0x%x)", pSelf, pTimeStamp, pIt);
+	cellPamf.notice("cellPamfReaderGetEpIteratorWithTimeStamp(pSelf=*0x%x, pTimeStamp=*0x%x, pIt=*0x%x)", pSelf, pTimeStamp, pIt);
+
+	if (pSelf->attribute & CELL_PAMF_ATTRIBUTE_MINIMUM_HEADER)
+	{
+		return CELL_PAMF_ERROR_NOT_AVAILABLE;
+	}
+
+	const u32 ep_num = pSelf->isPsmf ? vm::static_ptr_cast<const PsmfStreamHeader>(pSelf->currentStream)->ep_num : vm::static_ptr_cast<const PamfStreamHeader>(pSelf->currentStream)->ep_num;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1387]: use cellPamfReaderGetNumberOfEp

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 2]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
-	// TODO
 	return CELL_OK;
 }
 
 error_code cellPamfReaderGetEpIteratorWithTimeStamp(vm::ptr<CellPamfReader> pSelf, vm::ptr<CellCodecTimeStamp> pTimeStamp, vm::ptr<CellPamfEpIterator> pIt)
 {
-	cellPamf.todo("cellPamfReaderGetEpIteratorWithTimeStamp(pSelf=*0x%x, pTimeStamp=*0x%x, pIt=*0x%x)", pSelf, pTimeStamp, pIt);
+	cellPamf.notice("cellPamfReaderGetEpIteratorWithTimeStamp(pSelf=*0x%x, pTimeStamp=*0x%x, pIt=*0x%x)", pSelf, pTimeStamp, pIt);
+
+	if (pSelf->attribute & CELL_PAMF_ATTRIBUTE_MINIMUM_HEADER)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1382]: Missing sanity check for pSelf

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 3]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+	{
+		return CELL_PAMF_ERROR_NOT_AVAILABLE;
+	}
+
+	const u32 ep_num = pSelf->isPsmf ? vm::static_ptr_cast<const PsmfStreamHeader>(pSelf->currentStream)->ep_num : vm::static_ptr_cast<const PamfStreamHeader>(pSelf->currentStream)->ep_num;
+
+	pIt->num = ep_num;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1389]: missing sanity check for pIt

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 4]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+	pIt->isPamf = !pSelf->isPsmf;
+
+	if (ep_num == 0)
+	{
+		return CELL_PAMF_ERROR_EP_NOT_FOUND;
+	}
+
+	u32 i = ep_num - 1;
+	const u64 requested_time_stamp = std::bit_cast<be_t<u64>>(*pTimeStamp);
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1398]: missing sanity check for pTimeStamp

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 5]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
 
 error_code cellPamfEpIteratorGetEp(vm::ptr<CellPamfEpIterator> pIt, vm::ptr<CellPamfEp> pEp)
 {
-	cellPamf.todo("cellPamfEpIteratorGetEp(pIt=*0x%x, pEp=*0x%x)", pIt, pEp);
+	cellPamf.notice("cellPamfEpIteratorGetEp(pIt=*0x%x, pEp=*0x%x)", pIt, pEp);
 
 	// always returns CELL_OK
-	// TODO
+
+	if (pIt->isPamf)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1444]: Missing sanity check for pIt

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 6]
```diff
@@ -35,404 +36,996 @@ void fmt_class_string<CellPamfError>::format(std::string& out, u64 arg)
...
+	return CELL_OK;
 }
 
-u8 pamfGetStreamChannel(vm::ptr<CellPamfReader> pSelf, u32 stream)
+void pamfEpUnpack(vm::cptr<PamfEpHeader> ep_packed, vm::ptr<CellPamfEp> ep)
 {
-	// TODO: get stream channel correctly
-	ensure(stream < pSelf->pAddr->stream_count);
-	auto& header = pSelf->pAddr->stream_headers[stream];
+	ep->indexN = (ep_packed->value0 >> 14) + 1;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:304]: missing sanity checks for ep_packed and ep

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 7]
```diff
@@ -35,404 +36,996 @@ void fmt_class_string<CellPamfError>::format(std::string& out, u64 arg)
...
+	ep->nThRefPictureOffset = ((ep_packed->value0 & 0x1fff) * 0x800) + 0x800;
+	ep->pts.upper = ep_packed->pts_high;
+	ep->pts.lower = ep_packed->pts_low;
+	ep->rpnOffset = ep_packed->rpnOffset * 0x800ull;
+}
 
-	switch (header.type)
+void psmfEpUnpack(vm::cptr<PsmfEpHeader> ep_packed, vm::ptr<CellPamfEp> ep)
+{
+	ep->indexN = (ep_packed->value0 >> 14) + 1;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:313]: missing sanity checks for ep_packed and ep

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 8]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+
 	return CELL_OK;
 }
 
 s32 cellPamfEpIteratorMove(vm::ptr<CellPamfEpIterator> pIt, s32 steps, vm::ptr<CellPamfEp> pEp)
 {
-	cellPamf.todo("cellPamfEpIteratorMove(pIt=*0x%x, steps=%d, pEp=*0x%x)", pIt, steps, pEp);
+	cellPamf.notice("cellPamfEpIteratorMove(pIt=*0x%x, steps=%d, pEp=*0x%x)", pIt, steps, pEp);
+
+	u32 new_index = pIt->index + steps;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1460]: missing sanity check for pIt

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 9]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+
+	if (static_cast<s32>(new_index) < 0)
+	{
+		steps = -static_cast<s32>(pIt->index);
+		new_index = 0;
+	}
+	else if (new_index >= pIt->num)
+	{
+		steps = pIt->num - 1 - pIt->index;
+		new_index = pIt->num - 1;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1470]: I think this is:
```suggestion
		steps = (pIt->num - pIt->index)  - 1;
		new_index = pIt->index + steps;
```

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 10]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+	const u32 ep_num = pSelf->isPsmf ? vm::static_ptr_cast<const PsmfStreamHeader>(pSelf->currentStream)->ep_num : vm::static_ptr_cast<const PamfStreamHeader>(pSelf->currentStream)->ep_num;
+	u64 next_rpn_offset = pSelf->dataSize;
+
+	if (ep_num == 0)
+	{
+		return CELL_PAMF_ERROR_EP_NOT_FOUND;
+	}
+
+	u32 i = ep_num - 1;
+	const u64 requested_time_stamp = std::bit_cast<be_t<u64>>(*pTimeStamp);
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1517]: missing sanity check for pTimeStamp

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 11]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+	cellPamf.notice("cellPamfReaderGetEpWithTimeStamp(pSelf=*0x%x, pTimeStamp=*0x%x, pEp=*0x%x, unk=0x%x)", pSelf, pTimeStamp, pEp, unk);
+
+	// This function is broken on LLE
+
+	if (pSelf->attribute & CELL_PAMF_ATTRIBUTE_MINIMUM_HEADER)
+	{
+		return CELL_PAMF_ERROR_NOT_AVAILABLE;
+	}
+
+	const u32 ep_num = pSelf->isPsmf ? vm::static_ptr_cast<const PsmfStreamHeader>(pSelf->currentStream)->ep_num : vm::static_ptr_cast<const PamfStreamHeader>(pSelf->currentStream)->ep_num;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1508]: use cellPamfReaderGetNumberOfEp

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 12]
```diff
@@ -722,44 +1345,274 @@ error_code cellPamfReaderGetStreamInfo(vm::ptr<CellPamfReader> pSelf, vm::ptr<vo
...
+	u64 next_rpn_offset = pSelf->dataSize;
+
+	if (epIndex >= ep_num)
+	{
+		return CELL_PAMF_ERROR_INVALID_ARG;
+	}
+
+	if (pSelf->attribute & CELL_PAMF_ATTRIBUTE_MINIMUM_HEADER)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:1582]: missing sanity check for pSelf

**Review by Megamouse — COMMENTED** [review]
I haven't really reversed it 1:1 to check every single line of code.
For example I only flew over the verify function. But given that the rest seems mostly spot on, I'm confident that the verify doesn't have any pressing issues.
Good job. I think this can be merged when the review fixes are in.

Please keep in mind to push further changes to new commits, otherwise the re-review will be really annoying (unless i keep the branch around and rebase mixed).

File: rpcs3/Emu/Cell/Modules/cellPamf.h [hunk 13]
```diff
@@ -153,6 +170,7 @@ struct CellPamfEp
 	be_t<u32> nThRefPictureOffset;
 	CellCodecTimeStamp pts;
 	be_t<u64> rpnOffset;
+	be_t<u64> nextRpnOffset; // Only used in two undocumented functions
```
[the remark this answers is not in the document]
Comment (elad335) [reply rpcs3/Emu/Cell/Modules/cellPamf.h:173]: add nother struct then

File: rpcs3/Emu/Cell/Modules/cellPamf.h [hunk 14]
```diff
@@ -32,6 +32,23 @@ enum CellPamfStreamType
 	CELL_PAMF_STREAM_TYPE_USER_DATA  = 5,
 	CELL_PAMF_STREAM_TYPE_VIDEO      = 20,
 	CELL_PAMF_STREAM_TYPE_AUDIO      = 21,
+
+	CELL_PAMF_STREAM_TYPE_PSMF_AVC        = 6,
+	CELL_PAMF_STREAM_TYPE_PSMF_ATRAC3PLUS = 7,
+	CELL_PAMF_STREAM_TYPE_PSMF_LPCM       = 8,
+	CELL_PAMF_STREAM_TYPE_PSMF_USER_DATA  = 9,
+	CELL_PAMF_STREAM_TYPE_UNK             = 22,
```
[the remark this answers is not in the document]
Comment (capriots) [reply rpcs3/Emu/Cell/Modules/cellPamf.h:40]: CELL_PAMF_STREAM_TYPE_UNK is from the function at 0x9c (roughly equivalent to my pamfIsSameStreamType()), the others can be found in the functions at 0x2d0 (pamfStreamToTypeChannel()), 0x114 (pamfTypeChannelToStream()), 0x1798 (pamfVerify(), inside a switch, though LLE is just calling a bunch of functions that simply return 0 so I didn't bother with that) and cellPamfReaderGetStreamInfo().

File: rpcs3/Emu/Cell/Modules/cellPamf.cpp [hunk 15]
```diff
@@ -35,404 +36,996 @@ void fmt_class_string<CellPamfError>::format(std::string& out, u64 arg)
...
 }
 
 u8 cellPamfReaderGetNumberOfStreams(vm::ptr<CellPamfReader> pSelf)
 {
-	cellPamf.warning("cellPamfReaderGetNumberOfStreams(pSelf=*0x%x)", pSelf);
+	cellPamf.notice("cellPamfReaderGetNumberOfStreams(pSelf=*0x%x)", pSelf);
 
-	// cannot return error code
-	return pSelf->pAddr->stream_count;
+	return pSelf->pamf.currentGroup->stream_num;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellPamf.cpp:842]: still missing the sanity check

**Pull request merged**